src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts

```diff
@@ -155,19 +155,24 @@ export class NativeLifecycleService extends AbstractLifecycleService {
 
 	protected async handleWillShutdown(reason: ShutdownReason): Promise<void> {
 		const joiners: Promise<void>[] = [];
+		const lastJoiners: (() => Promise<void>)[] = [];
 		const pendingJoiners = new Set<IWillShutdownEventJoiner>();
 		const cts = new CancellationTokenSource();
-
 		this._onWillShutdown.fire({
 			reason,
 			token: cts.token,
 			joiners: () => Array.from(pendingJoiners.values()),
-			join(promise, joiner) {
-				joiners.push(promise);
-
-				// Track promise completion
+			join(promiseOrPromiseFn, joiner) {
 				pendingJoiners.add(joiner);
-				promise.finally(() => pendingJoiners.delete(joiner));
+
+				if (joiner.order === WillShutdownJoinerOrder.Last) {
+					const promiseFn = typeof promiseOrPromiseFn === 'function' ? promiseOrPromiseFn : () => promiseOrPromiseFn;
+					lastJoiners.push(() => promiseFn().finally(() => pendingJoiners.delete(joiner)));
+				} else {
+					const promise = typeof promiseOrPromiseFn === 'function' ? promiseOrPromiseFn() : promiseOrPromiseFn;
+					promise.finally(() => pendingJoiners.delete(joiner));
+					joiners.push(promise);
+				}
```

Comment:
Please let me know if this still makes sense to you, I felt it was easier to understand.